File: docs/example/collection/vec/Vec.cs
module collection.vec

type vec_t(A: type_t, length: Nat) {
  null_vec_t(length: Zero)
  cons_vec_t(n: Nat, length: Succ(n), head: A, tail: Vec(A, n))
}

@infix(++)
vec_append(ante: Vec(A, m), succ: Vec(A, n)): Vec(A, m + n) =
  ante case {
    NullVec => succ
    ConsVec => ConsVec(
      head = ante.head,
      tail = ante.tail ++ succ)
  }

vec_map(fun: A -> B, vec: Vec(A, n)): Vec(A, n) =
  vec case {
    NullVec => vec
    ConsVec => ConsVec(
      head = fun(vec.head),
      tail = vec_map(fun, vec.tail))
  }

File: src/main/scala/xieyuheng/cicada/prelude/vec.scala
package xieyuheng.cicada.prelude

import xieyuheng.cicada._
import xieyuheng.cicada.expDSL._
import xieyuheng.cicada.pretty._

object vec {

  val env = Env()

  .import_all(nat.env)

  .define_type("vec_t", %(
    "A" -> the_type,
    "length" -> the("nat_t")),
    members = %(
      "null_vec_t" -> %(
        "length" -> "zero_t"),
      "cons_vec_t" -> %(
        "n" -> the("nat_t"),
        "length" -> ("succ_t" ap %("prev" -> "n")),
        "head" -> the("A"),
        "tail" -> the("vec_t" ap %("A" -> "A", "length" -> "n")))))

}
